### ISSUE TITLE
QR code scanning for windows desktop pairing doesn't work with Samsung A21

### ISSUE THREAD

**Issue by twmillett** [signalapp/Signal-Android#10756]
- [X] I have searched open and closed issues for duplicates
-- it's https://github.com/signalapp/Signal-Android/issues/4637, which was closed in a cleanup sweep without a fix
-- https://github.com/signalapp/Signal-Android/pull/5538 is related, but not merged
-- various interweb threads on this: https://www.reddit.com/r/signal/comments/l0v727/signal_desktop_qr_code_not_scanning_on_my_android/

- [X] I am submitting a bug report for existing functionality that does not work as intended
- [X] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [X] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description

On Samsung A21, Android app's QR scanner does not recognize the QR code generated by the Windows desktop app

### Steps to reproduce

Actual
- Install signal from the Play store on Samsung A21
- Install signal for Windows
- Open the Windows app & be presented with a QR code
- Open Signal on Samsung A21 -> settings -> linked devices -> (+)
- Observe the screen inviting you to "scan the code displayed on the device to link"
- Position the camera widget over the QR code

Expected
- The phone recognizes the QR code

Actual
- Nothing happens.  The phone app stays on the screen, the screen eventually dims, etc

Attempts:
- I read various threads about screen shotting the code, zooming, adjusting screen brightness, etc.  Same thing.  I moved the QR code to another screen, via thumdrive.  Same thing.
- scanning the QR code via the Microsoft launcher / QR reader IMMEDIATELY recognizes the QR code and searches the web for a tsdevice URL, which isn't found of course

### Screenshots
N/A
### Device info
**Device:** Samsung A21
**Android version:** 10
**Signal version:** 5.2.3

### Link to debug log
https://debuglogs.org/07cc6e90f356a758ed7757d79ff499e6394c04177efaa9825ea497cc9ac6f004

**Comment by Caerbannog** [comment]
Hi @twmillett !

I have looked at your problem some time, and can't find a clear answer yet. Note: I'm not a member of the Signal team.
Here are my findings, maybe someone else will know more. And some questions to troubleshoot your problem.

1. Your log contains a lot of errors from the QR-code scanning library (zxing).
Lots of `com.google.zxing.FormatException`, and a few `com.google.zxing.ChecksumException`. And lots of `01-21 12:24:39.213 31644 12401 W ScanningThread: null` that I don't understand (what line logged them?).
The first thing should do is check that this QR-code is valid, because there is a small chance that the bug is in Signal-dekstop.
    **Use any QR-code app and paste the result here**, hiding important information (though I *think* it's not a security risk).
Example from my own Signal-desktop: `tsdevice:/?uuid=xxxxxxxxxxxxxxxxxxxxxxxxx&pub_key=xxxxxxxxxxxxxxxxxxxxxxxxx`.

2. Could you try to scan a random valid QR code? It should be detected, the message "Do want to link this device?" should appear, and then a toast should say "Sorry, this is not a valid code".
    **Scan this then describe what happens:** https://en.wikipedia.org/wiki/QR_code#/media/File:QR_code_for_mobile_English_Wikipedia.svg

3. I read that in some cases a zxing variant worked better on paper. Or might have had problems with rotations. ([source](https://github.com/Redth/ZXing.Net.Mobile/issues/16))
    **Try all phone rotations (90°, 180°). Then print the QR code.** 

4. Your debug log was useful, but I think that it misses the app startup part because you tried multiple times before extracting the log.
     **Kill the app, then restart it, try to scan the QR-code for 2 seconds, then extract a debug log quickly**

I hope we'll find the solution to get you started, it's a great app :wink: 
I you find a solution please give as much details as possible here to help other people.

**Comment by twmillett** [comment]
TL;DR

Printing out the QR code from MS Paint totally worked!  FWIW, I printing mine out such that the image was fit to the page (so, about 8 inches square on a standard 8.5x11 page.  After scanning successfully, my machine was able to pair, no problem.

Thank you so much for taking the time to respond and, even, solve the issue for me!

> Hi @twmillett !
> 
> I have looked at your problem some time, and can't find a clear answer yet. Note: I'm not a member of the Signal team.
> Here are my findings, maybe someone else will know more. And some questions to troubleshoot your problem.
> 
> 1. Can you confirm that the app shows the camera preview in a rectangle, like any other QR-code app? Or do you see nothing?

Yes, I can confirm that I see the camera preview in a rectangle.  I've attached a screen shot of me trying to scan the wikipedia page code you mentioned.

![image](https://user-images.githubusercontent.com/8038952/105421300-c3345000-5bf6-11eb-8e96-b4f3ce1d9f94.png)

> 2. Your log contains a lot of errors from the QR-code scanning library (zxing).
>    Lots of `com.google.zxing.FormatException`, and a few `com.google.zxing.ChecksumException`. And lots of `01-21 12:24:39.213 31644 12401 W ScanningThread: null` that I don't understand (what line logged them?).
>    The first thing should do is check that this QR-code is valid, because there is a small chance that the bug is in Signal-dekstop.
>    **Use any QR-code app and paste the result here**, hiding important information (though I _think_ it's not a security risk).
>    Example from my own Signal-desktop: `tsdevice:/?uuid=xxxxxxxxxxxxxxxxxxxxxxxxx&pub_key=xxxxxxxxxxxxxxxxxxxxxxxxx`.

Scanning from the Microsoft launcher app's QR code reader gives me this url

tsdevice:/?uuid=xxxxxxxxxxxxaLw&pub_key=xxxxxxxxxxcZ

(I replaced a bunch of stuff with xxx's)

> 3. Could you try to scan a random valid QR code? It should be detected, the message "Do want to link this device?" should appear, and then a toast should say "Sorry, this is not a valid code".
>    **Scan this then describe what happens:** https://en.wikipedia.org/wiki/QR_code#/media/File:QR_code_for_mobile_English_Wikipedia.svg

Same behavior as with the device code -- it doesn't appear to recognize the existence of the QR code.

> 4. I read that in some cases a zxing variant worked better on paper. Or might have had problems with rotations. ([source](https://github.com/Redth/ZXing.Net.Mobile/issues/16))
>    **Try all phone rotations (90°, 180°). Then print the QR code.**

Using all different rotations didn't help...BUT....as mentioned above

- taking a screenshot of the code in Windows
- pasting it in paint (still won't scan from here, even enlarged to full screen)
- printing it out on an 8.5"x11" piece of paper it did scan!  yay!

> 5. Your debug log was useful, but I think that it misses the app startup part because you tried multiple times before extracting the log.
>    **Kill the app, then restart it, try to scan the QR-code for 2 seconds, then extract a debug log quickly**
> 
> I hope we'll find the solution to get you started, it's a great app 😉
> I you find a solution please give as much details as possible here to help other people.

The log is here, FWIW.  But, again, printing out the code totally worked!

https://debuglogs.org/529828c10e2b49f4deb38121265fd33c8b6ca87fe57d81d130565397f6bb95ca

**Comment by bannisdale** [comment]
In search for a solution, I came here. As suggested, I tried scanning the QR code with my inbuilt camera. While I tried all angles, distances and sizes (via screenshot) with the Signal app, the QR code got read instantaniously by my camera app. However, then there was a message that asked me to do the scan within the signal app for security reasons.

That is to say: The inbuilt QR scanner is – sorry – garbage. When I link Threema, the phone hasn't even come to a stable position when the connection request is started. With Signal, it seems like you'll have to print (what?!) the QR code even though it's obviously totally readable for common QR scanners. That's way below subpar. :(

**Comment by cody-signal** [comment]
Can you try the latest beta? We've done some rework on it.

**Comment by bannisdale** [comment]
@cody-signal Thanks for the quick reply! Great timing. :) 

I enrolled in the beta in the Play Store and updated the app today. I'm not sure if it really is a beta version, but for reference, client version is showing 5.46.5 now.

Anyhow, the QR code was read/recognized almost instantaniously. Unlinked the devices, linked again, same behavior. It now works as seemlessly (even a bit better) as it would be expected now. At least for me.

Good work! :)

(I haven't updated the Desktop client in the meantime, so it seems that it's indeed an improvement in the app.)
